subsys/bluetooth/services/ras/rrsp/ras_rrsp.c

```diff
@@ -8,21 +8,671 @@
+	k_work_init(&rrsp->status_work, status_work_handler);
+	k_timer_init(&rrsp->rascp_timeout, rascp_timeout_handler, NULL);
+
+	return 0;
+}
+
+void bt_ras_rrsp_free(struct bt_conn *conn)
+{
+	struct bt_ras_rrsp *rrsp = rrsp_find(conn);
+
+	if (rrsp) {
+		LOG_DBG("conn %p rrsp %p", (void *)rrsp->conn, (void *)rrsp);
+
+		(void)k_work_cancel(&rrsp->send_data_work);
+		(void)k_work_cancel(&rrsp->rascp_work);
+		(void)k_work_cancel(&rrsp->status_work);
+		k_timer_stop(&rrsp->rascp_timeout);
+
+		k_work_queue_drain(&rrsp_wq, false);
+
+		bt_conn_unref(rrsp->conn);
+		rrsp->conn = NULL;
+	}
+}
+
+static ssize_t ras_features_read(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+				 uint16_t len, uint16_t offset)
 {
 	return bt_gatt_attr_read(conn, attr, buf, len, offset, &ras_features, sizeof(ras_features));
 }
 
+static ssize_t rd_ready_read(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			     uint16_t len, uint16_t offset)
+{
+	struct bt_ras_rrsp *rrsp = rrsp_find(conn);
+
+	__ASSERT_NO_MSG(rrsp);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &rrsp->ready_ranging_counter,
+				 sizeof(uint16_t));
+}
+
+static ssize_t rd_overwritten_read(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+				   uint16_t len, uint16_t offset)
+{
+	struct bt_ras_rrsp *rrsp = rrsp_find(conn);
+
+	__ASSERT_NO_MSG(rrsp);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &rrsp->overwritten_ranging_counter,
+				 sizeof(uint16_t));
+}
+
+static ssize_t ras_cp_write(struct bt_conn *conn, struct bt_gatt_attr const *attr, void const *buf,
+			    uint16_t len, uint16_t offset, uint8_t flags)
+{
+	LOG_HEXDUMP_DBG(buf, len, "Write request");
+
+	if (!bt_gatt_is_subscribed(conn, attr, BT_GATT_CCC_INDICATE)) {
+		LOG_DBG("Not subscribed");
+		return BT_GATT_ERR(RAS_ATT_ERROR_CCC_CONFIG);
+	}
+
+	struct bt_ras_rrsp *rrsp = rrsp_find(conn);
+
+	__ASSERT_NO_MSG(rrsp);
+
+	if (k_work_is_pending(&rrsp->rascp_work) || len > RASCP_WRITE_MAX_LEN) {
+		LOG_DBG("Write rejected");
+		return BT_GATT_ERR(RAS_ATT_ERROR_WRITE_REQ_REJECTED);
+	}
+
+	uint16_t buf_len = MIN(len, RASCP_WRITE_MAX_LEN);
+
+	memcpy(rrsp->rascp_cmd_buf, buf, buf_len);
+	rrsp->rascp_cmd_len = (uint8_t)buf_len;
+
+	k_work_submit_to_queue(&rrsp_wq, &rrsp->rascp_work);
+
+	return len;
+}
+
+static void ondemand_rd_ccc_cfg_changed(struct bt_gatt_attr const *attr, uint16_t value)
+{
+	LOG_DBG("On-demand Ranging Data CCCD changed: %u", value);
+}
+
+static void ras_cp_ccc_cfg_changed(struct bt_gatt_attr const *attr, uint16_t value)
+{
+	LOG_DBG("RAS-CP CCCD changed: %u", value);
+}
+
+static void rd_ready_ccc_cfg_changed(struct bt_gatt_attr const *attr, uint16_t value)
+{
+	LOG_DBG("Ranging Data Ready CCCD changed: %u", value);
+}
+
+static void rd_overwritten_ccc_cfg_changed(struct bt_gatt_attr const *attr, uint16_t value)
+{
+	LOG_DBG("Ranging Data Overwritten CCCD changed: %u", value);
+}
+
+#if defined(CONFIG_BT_RAS_RRSP_AUTO_ALLOC_INSTANCE)
```

Comment:
Ensure if a link doesn't have a RRSP allocated things fail gracefully.
Need to add some tests for this.